tensorflow/core/kernels/training_ops.h

```diff
@@ -92,6 +92,17 @@ struct ApplyAdagradDA {
                   typename TTypes<T>::ConstFlat grad);
 };
 
+template <typename Device, typename T, typename Tindex, bool has_epsilon>
+struct SparseApplyAdagrad {
+  Status operator()(const Device& d, typename TTypes<T>::Matrix var,
+                    typename TTypes<T>::Matrix accum,
+                    typename TTypes<T>::ConstScalar lr,
+                    typename TTypes<T>::ConstScalar epsilon,
```

Comment:
Please document what is the semantic of `epsilon` if `has_epsilon` is false (it is ignored right?).

Reply:
Done.